R/tmb-methods.R

```diff
@@ -207,3 +207,111 @@ print.mmrm_tmb <- function(x,
+#' and this is the only type available for use with models with a spatial covariance structure.
+#' @param ... \cr not used in this method.
+#' @return Vector of residuals
+#' @importFrom stats residuals
+#' @exportS3Method
+#' @examples
+#' # residuals:
+#' residuals(object, type = "response")
+#' residuals(object, type = "pearson")
+#' residuals(object, type = "normalized")
+#' @references
+#' - \insertRef{galecki2013linear}{mmrm}
+residuals.mmrm_tmb <- function(object, type = c("response", "pearson", "normalized"), ...) {
+  assert_class(object, "mmrm_tmb")
+  assert_character(type)
+  type <- match.arg(type, several.ok = FALSE)
+  resids_unscaled <- component(object, "y_vector") - fitted(object)
+  if (type == "response") {
+    return(unname(resids_unscaled))
+  } else {
+    if (object$formula_parts$is_spatial) {
+      stop("Only raw / response residuals are available for models with spatial covariance structures.")
+    }
+    if (type == "pearson") {
+      h_residuals_pearson(object, resids_unscaled)
+    } else if (type == "normalized") {
+      h_residuals_normalized(object, resids_unscaled)
+    }
+  }
+}
+
+#' Calculate Pearson residuals
+#'
+#' This is used by [residuals.mmrm_tmb()] to calculate Pearson residuals.
+#'
+#' @param object (`mmrm_tmb`)\cr the fitted MMRM.
+#' @param resids_unscaled (`numeric`)\cr the raw/response residuals.
+#'
+#' @return Vector of residuals
+#'
+#' @keywords internal
+h_residuals_pearson <- function(object, resids_unscaled) {
+  visits <- as.numeric(object$tmb_data$full_frame[[object$formula_parts$visit_var]])
+  if (component(object, "n_groups") == 1) {
+    visit_sigmas <- sqrt(diag(object$cov))
+    resids <- resids_unscaled / visit_sigmas[visits] * sqrt(object$tmb_data$weights_vector)
+  } else {
+    grp_visit_sigmas <- lapply(object$cov, function(x) sqrt(diag(x)))
+    subject_grps <- object$tmb_data$full_frame[[object$formula_parts$group_var]]
+    nobs <- nrow(object$tmb_data$full_frame)
+    resids <- sapply(1:nobs, function(x) {
+      resids_unscaled[x] / grp_visit_sigmas[[subject_grps[x]]][visits[x]] * sqrt(object$tmb_data$weights_vector[x])
+    })
+  }
+  return(unname(resids))
+}
+
+#' Calculate normalized residuals
+#'
+#' This is used by [residuals.mmrm_tmb()] to calculate normalized / scaled residuals.
+#'
+#' @param object (`mmrm_tmb`)\cr the fitted MMRM.
+#' @param resids_unscaled (`numeric`)\cr the raw/response residuals.
+#'
+#' @return Vector of residuals
+#'
+#' @keywords internal
+h_residuals_normalized <- function(object, resids_unscaled) {
+  resid_df <- data.frame(subject = object$tmb_data$full_frame[[object$formula_parts$subject_var]],
+                         time = as.numeric(object$tmb_data$full_frame[[object$formula_parts$visit_var]]),
+                         residual = resids_unscaled,
+                         weights = object$tmb_data$weights_vector)
+
+  subject_list <- split(resid_df, resid_df$subject)
+
+  if (component(object, "n_groups") == 1) {
```

Comment:
```suggestion
  lower_chol_list <- if (component(object, "n_groups") == 1) {
```
and then remove inner assignment within curly brackets